lib/src/main/java/com/exallium/rxrecyclerview/lib/RxRecyclerViewAdapter.java

```diff
@@ -132,20 +134,31 @@ public void onError(Throwable e) {
         @Override
         public void onNext(EventElement<K, V> rxEvent) {
             preProcessElement(rxEvent);
+            EventElement<K, V> currentRxEvent = null;
             switch (rxEvent.getData().getType()) {
                 case ADD:
-                    if (treeSet.add(rxEvent)) {
-                        notifyItemInserted(getIndexOf(rxEvent));
-                    } else {
-                        treeSet.remove(rxEvent);
+                    if((currentRxEvent = mItems.put(rxEvent.getData().getKey(), rxEvent)) != null){
+                        final int orgPos = getIndexOf(currentRxEvent);
+                        treeSet.remove(currentRxEvent);
                         treeSet.add(rxEvent);
-                        notifyItemChanged(getIndexOf(rxEvent));
+                        final int newPos = getIndexOf(rxEvent);
+                        if(orgPos != newPos){
+                            notifyItemMoved(orgPos, newPos);
+                        }
+                        notifyItemChanged(newPos);
+                    }
+                    else {
+                        treeSet.add(rxEvent);
+                        notifyItemInserted(getIndexOf(rxEvent));
                     }
                     break;
                 case REMOVE:
-                    int index = getIndexOf(rxEvent);
-                    if (treeSet.remove(rxEvent)) {
-                        notifyItemRemoved(index);
+                    currentRxEvent = mItems.remove(rxEvent.getData().getKey());
+                    if(currentRxEvent != null){
+                        int index = getIndexOf(currentRxEvent);
```

Comment:
Move to inside of if statement, and combine the if statements.  ```getIndexOf``` is currently O(n), so I'd rather avoid calling it if unnecessary.